src/libraries/System.Linq/tests/ChunkTests.cs

```diff
@@ -0,0 +1,159 @@
+            Assert.Equal(new[] {9999, 0, 888}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {-1, 66, -777}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {1, 2, -12345}, chunks.Current);
+            Assert.False(chunks.MoveNext());
+        }
+
+        [Theory]
+        [InlineData(new[] {9999, 0, 888, -1, 66, -777, 1, 2}, typeof(TestReadOnlyCollection<int>))]
+        [InlineData(new[] {9999, 0, 888, -1, 66, -777, 1, 2}, typeof(TestCollection<int>))]
+        [InlineData(new[] {9999, 0, 888, -1, 66, -777, 1, 2}, typeof(TestEnumerable<int>))]
+        public void ChunkSourceUnevenly(int[] array, Type type)
+        {
+            IEnumerable<int> source = ConvertToType(array, type);
+
+            using IEnumerator<int[]> chunks = source.Chunk(3).GetEnumerator();
+            chunks.MoveNext();
+            Assert.Equal(new[] {9999, 0, 888}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {-1, 66, -777}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {1, 2}, chunks.Current);
+            Assert.False(chunks.MoveNext());
+        }
+
+        [Theory]
+        [InlineData(new[] {9999, 0}, typeof(TestReadOnlyCollection<int>))]
+        [InlineData(new[] {9999, 0}, typeof(TestCollection<int>))]
+        [InlineData(new[] {9999, 0}, typeof(TestEnumerable<int>))]
+        public void ChunkSourceSmallerThanMaxSize(int[] array, Type type)
+        {
+            IEnumerable<int> source = ConvertToType(array, type);
+
+            using IEnumerator<int[]> chunks = source.Chunk(3).GetEnumerator();
+            chunks.MoveNext();
+            Assert.Equal(new[] {9999, 0}, chunks.Current);
+            Assert.False(chunks.MoveNext());
+        }
+
+        [Theory]
+        [InlineData(new int[] {}, typeof(TestReadOnlyCollection<int>))]
+        [InlineData(new int[] {}, typeof(TestCollection<int>))]
+        [InlineData(new int[] {}, typeof(TestEnumerable<int>))]
+        public void EmptySourceYieldsNoChunks(int[] array, Type type)
+        {
+            IEnumerable<int> source = ConvertToType(array, type);
+
+            using IEnumerator<int[]> chunks = source.Chunk(3).GetEnumerator();
+            Assert.False(chunks.MoveNext());
+        }
+
+        [Fact]
+        public void RemovingFromSourceBeforeIterating()
+        {
+            var list = new List<int>
+            {
+                9999, 0, 888, -1, 66, -777, 1, 2, -12345
+            };
+            using IEnumerator<int[]> chunks = list.Chunk(3).GetEnumerator();
+            list.Remove(66);
+
+            chunks.MoveNext();
+            Assert.Equal(new[] {9999, 0, 888}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {-1, -777, 1}, chunks.Current);
+            chunks.MoveNext();
+            Assert.Equal(new[] {2, -12345}, chunks.Current);
+            Assert.False(chunks.MoveNext());
```

Comment:
I think we should be fine with validating a weaker property:

```suggestion
            IEnumerable<int[]> chunks = list.Chunk(3);
            list.Remove(66);
            
            Assert.Equal(new[] {new[] {9999, 0, 888}, new[] {-1, -777, 1}, new[] {2, -12345}}, chunks);
```
Modifying the source while enumerating is undefined behaviour, and could often result in exceptions.